app/versioning.py

```diff
@@ -0,0 +1,62 @@
+from typing import Callable, List
+
+import flask
+from werkzeug.exceptions import BadRequest
+
+VALID_API_VERSIONS = ['1.0']
+"""API versions that are valid for a user to request.
+
+NOTE: the first value in this list should be the latest supported version, but the
+ordering of subsequent version numbers is not important.
+"""
+LATEST_API_VERSION = VALID_API_VERSIONS[0]
+"""latest API version to default to when none other is requested"""
+
+
+def versioned(valid_versions: List[str] = None, throw_on_invalid: bool = True):
+    """
+    Extracts requested API version from X-API-Version HTTP header and passes
+    to wrapped function in the `version` kwarg.
+
+    :param throw_on_invalid: whether to throw exception on invalid passed version,
+        defaults to `True`
+    :param valid_versions: list of API versions to consider as valid, falls back to
+        default list of valid API versions when not specified
+    :raise InvalidApiVersion: when `throw_on_invalid` is `True` and invalid API version
+        was specified in the client request
+    """
+
+    def decorator(function: Callable):
+        def wrapper(*args, **kwargs):
+            if flask.has_request_context():
+                version = flask.request.headers.get('x-api-version', LATEST_API_VERSION)
+                if version not in valid_versions:
+                    if throw_on_invalid:
+                        raise InvalidApiVersion(version)
+                    else:
+                        version = LATEST_API_VERSION
+            else:
+                version = LATEST_API_VERSION
+
+            kwargs['version'] = float(version)
+            return function(*args, **kwargs)
+
+        return wrapper
+
+    # When decorator is not called as a function
+    if callable(valid_versions):
+        wrapped_function = valid_versions
+        valid_versions = VALID_API_VERSIONS
+        return decorator(wrapped_function)
+    else:
+        if not valid_versions:
+            valid_versions = VALID_API_VERSIONS
+        return decorator
+
+
+class InvalidApiVersion(BadRequest):
```

Comment:
Instead of returning a `400`, which is pretty generic, do you think we could return a `406 Not Acceptable`?

Reply:
I’m used to using 400 for user input errors. I believe not acceptable is specific to the accept header or maybe content type. If you’d like I can take a look of any other standard error could work, but most API users should be familiar with 400 being an input value error.

Reply:
I was always taught that 400 is what you use when there's not a more specific error that you can return. I think you're right about the use for 406. We could consider changing it to an `Accept` header rather than an `x-api-version` header. So you would send something like `Accept: application/1.0` in the request. Maybe that's overkill 🤷‍♂

Reply:
I've looked, and honestly, 400 should be the correct response code for this scenario. Everything else has to do with content types or request body content (besides other more strange scenarios). Unless you have a strong reason to do otherwise, I suggest we stick with a 400. I've engaged with a lot of third party APIs, and almost without exception, a 400 is standard for this particular type of scenario.

Reply:
Ok good enough for me. As long as it's a 4xx error it's semantically correct, so 400 is fine